copado/main/default/classes/CopadoDomain_Test.cls

```diff
@@ -0,0 +1,94 @@
+@IsTest
+public class CopadoDomain_Test {
+
+    @IsTest
+    private static void happyPath() {
+
+        // Setup
+        setup();
+
+        Credential crd1 = new Credential();
+        Credential crd2 = new Credential();
+
+
+        Environment env1 = new Environment().add(crd1);
+        Environment env2 = new Environment().add(crd2);
+
+        // Exercise
+
+        /*
+         * Ideally it should look like below:
+         *
+
+        new Pipeline()
+                .add( new Connection(env1, env2) )
+
+                .add( new Connection(env2,
+                        new Environment()
+                                .add( new Credential() )))
+
+                .add(  new Project()
+                                .add( new Release()
+                                                .add( new Promotion(crd1, crd2)
+                                                        .add( new Deployment()
+                                                                .add( new Step() )
+                                                                .add( new UserStory() ))
+                                                        )))
+
+                .persist();
+        */
+
+
+        Release rel = new Release();
+
+        new Pipeline()
+                .add( new Connection(env1, env2) )
+
+                .add( new Connection(env2,
+                        new Environment()
+                                .add( new Credential() )))
+
+                .add( new Project()
+                                .add( rel ));
+
+
+        Deployment dep = new Deployment();
+
+        new Promotion(rel, crd1, crd2)
+                .add( dep
+                        .add( new Step() ));
+
+        dep.add( new UserStory() ).persist();

```

Comment:
How it should look highly depends on what the actual Assert in the Test is interested in.

An example: if you want to test the bahavour of a Promotion a 2-3 lines with

Promotion pro = new Promotion();
...

should basically behin the scenes set up everything needed (Project, Pipeline, etc).

So to tell you if the implementation is good enough I would need to see an actual converted real life test.

Reply:
I understand, I will convert some real life test once I get this working